### ISSUE TITLE
"systemctl status" Regression when running libvirt VM

### ISSUE THREAD

**Issue by Toolybird** [systemd/systemd#22089]
**systemd version the issue has been seen with**
 > systemd 250 (250.2-2-arch)

**Used distribution**
 > Arch

**Linux kernel version used** (`uname -a`)

 > `Linux ryzen 5.15.13-arch1-1 #1 SMP PREEMPT Wed, 05 Jan 2022 16:20:59 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > normal output from "systemctl status"

**Unexpected behaviour you saw**
 > systemctl status
> Failed to dump process list for 'ryzen', ignoring: Operation not supported

**Steps to reproduce the problem**
 > fire up any libvirt VM
run systemctl status

Not sure if this is a libvirt or a systemd issue. It only started as of systemd-2.50.

Obvious suspect:
https://github.com/systemd/systemd/pull/21172

Arch Forum report:
https://bbs.archlinux.org/viewtopic.php?id=272962

Thanks

**Comment by mrc0mmand** [comment]
Can reproduce this locally with `systemd 250-3` as well.

```
[root@arch ~]# virsh list
 Id   Name           State
------------------------------
 1    root_default   running
[root@arch ~]# SYSTEMD_LOG_LEVEL=debug systemctl status --no-pager
Bus n/a: changing state UNSET → OPENING
sd-bus: starting bus by connecting to /run/systemd/private...
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.DBus.Properties member=GetAll cookie=1 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.systemd1 destination=n/a path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=a{sv} error-name=n/a error-message=n/a
● arch.localdomain
    State: running
     Jobs: 0 queued
   Failed: 0 units
    Since: Wed 2022-01-12 10:28:41 UTC; 1h 0min ago
   CGroup: /
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=GetUnitProcesses cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=error sender=org.freedesktop.systemd1 destination=n/a path=n/a interface=n/a member=n/a cookie=2 reply_cookie=2 signature=s error-name=org.freedesktop.DBus.Error.NotSupported error-message=Operation not supported
Failed to dump process list for 'arch.localdomain', ignoring: Operation not supported
Bus n/a: changing state RUNNING → CLOSED
```

**Comment by mrc0mmand** [comment]
Reproducer without `systemctl`:
```
[root@arch ~]# SYSTEMD_LOG_LEVEL=debug busctl call org.freedesktop.systemd1 /org/freedesktop/systemd1 org.freedesktop.systemd1.Manager GetUnitProcesses "s" -- "-.slice"
Bus n/a: changing state UNSET → OPENING
sd-bus: starting bus by connecting to /run/dbus/system_bus_socket...
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.120 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=GetUnitProcesses cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=error sender=:1.1 destination=:1.120 path=n/a interface=n/a member=n/a cookie=8414 reply_cookie=2 signature=s error-name=org.freedesktop.DBus.Error.NotSupported error-message=Operation not supported
Call failed: Operation not supported
Bus n/a: changing state RUNNING → CLOSED
```

**Comment by mrc0mmand** [comment]
So, the problematic commit appears to be https://github.com/systemd/systemd/pull/21172/commits/038cae098b6d85ad9fb06594d36f49915d32f000. And the root cause is that `GetUnitProcesses` is not a valid operation for the root slice (`-.slice`):

```
# busctl call org.freedesktop.systemd1 /org/freedesktop/systemd1 org.freedesktop.systemd1.Manager GetUnitProcesses "s" --  "-.slice"
Call failed: Operation not supported
# busctl call org.freedesktop.systemd1 /org/freedesktop/systemd1 org.freedesktop.systemd1.Manager GetUnitProcesses "s" --  "system.slice"
a(sus) 19 "/system.slice/systemd-networkd.service" 251 "/usr/lib/systemd/systemd-networkd" "/system.slice/systemd-udevd.service" 247 "/usr/lib/systemd/systemd-udevd" "/system.slice/polkit.service" 303 "/usr/lib/polkit-1/polkitd --no-debug" "/system.slice/libvirtd.service" 363 "/usr/bin/libvirtd --timeout 120" "/system.slice/libvirtd.service" 498 "/usr/bin/dnsmasq --conf-file=/var/lib/libvirt/dnsmasq/vagrant-libvirt.conf --leasefile-ro --dhcp-script=/usr/lib/libvirt/libvirt_leaseshelper" "/system.slice/libvirtd.service" 499 "/usr/bin/dnsmasq --conf-file=/var/lib/libvirt/dnsmasq/vagrant-libvirt.conf --leasefile-ro --dhcp-script=/usr/lib/libvirt/libvirt_leaseshelper" "/system.slice/systemd-journald.service" 243 "/usr/lib/systemd/systemd-journald" "/system.slice/sshd.service" 301 "\"sshd: /usr/bin/sshd -D [listener] 0 of 10-100 startups\"" "/system.slice/systemd-machined.service" 362 "/usr/lib/systemd/systemd-machined" "/system.slice/dhcpcd.service" 291 "\"dhcpcd: [manager] [ip4] [ip6]\"" "/system.slice/dhcpcd.service" 292 "\"dhcpcd: [privileged proxy]\"" "/system.slice/dhcpcd.service" 293 "\"dhcpcd: [network proxy]\"" "/system.slice/dhcpcd.service" 294 "\"dhcpcd: [control proxy]\"" "/system.slice/dhcpcd.service" 315 "\"dhcpcd: [BPF ARP] eth0 192.168.122.175\"" "/system.slice/systemd-resolved.service" 280 "/usr/lib/systemd/systemd-resolved" "/system.slice/dbus.service" 287 "/usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only" "/system.slice/system-getty.slice/getty@tty1.service" 304 "/sbin/agetty -o \"-p -- \\\\u\" --noclear - linux" "/system.slice/virtlogd.service" 509 "/usr/bin/virtlogd" "/system.slice/systemd-logind.service" 295 "/usr/lib/systemd/systemd-logind"
```

/cc @poettering @keszybz

**Comment by mrc0mmand** [comment]
Ah, it's not `-.slice`, but `machine.slice`:
```
# busctl call org.freedesktop.systemd1 /org/freedesktop/systemd1 org.freedesktop.systemd1.Manager GetUnitProcesses "s" --  "machine.slice"
Call failed: Operation not supported
```

**Comment by mrc0mmand** [comment]
A bit more, possibly interesting stuff:

```
# systemd-cgls -a  -u machine.slice
Unit machine.slice (/machine.slice):
└─machine-qemu\x2d3\x2drootdefault.scope … (#2876)
  → trusted.delegate: 1
  → trusted.invocation_id: d436daf25abd4597979168389893713b
  └─libvirt (#2928)
    ├─1507 /usr/bin/qemu-system-x86_64 -name guest=root_default,debug-threads=on -S -object {"qom-type":"secret","id":"masterKey0","format":"raw","file":"/var/lib/libvirt/qemu/domain-3-root_default/master-key>
    ├─vcpu1 (#3033)
    ├─vcpu0 (#3005)
    └─emulator (#2977)
```

```
# busctl call org.freedesktop.systemd1 /org/freedesktop/systemd1 org.freedesktop.systemd1.Manager GetUnitProcesses "s" --  "machine-qemu\x2d3\x2drootdefault.scope"
Call failed: Operation not supported
```

```
# systemctl cat "machine-qemu\x2d3\x2drootdefault.scope"
# /run/systemd/transient/machine-qemu\x2d3\x2drootdefault.scope
# This is a transient unit file, created programmatically via the systemd API. Do not edit.
[Scope]
Slice=machine.slice

[Unit]
Description=Virtual Machine qemu-3-rootdefault

[Scope]
Delegate=yes

[Unit]
CollectMode=inactive-or-failed

[Scope]
TasksMax=16384
Slice=machine.slice

[Unit]
After=libvirtd.service
Before=virt-guest-shutdown.target

```

**Comment by mrc0mmand** [comment]
And to be even more specific:

```
openat(AT_FDCWD, "/sys/fs/cgroup/machine.slice/machine-qemu\\x2d3\\x2drootdefault.scope/cgroup.procs", O_RDONLY|O_CLOEXEC) = 46
newfstatat(46, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(46, "", 4096)                      = 0
openat(AT_FDCWD, "/sys/fs/cgroup/machine.slice/machine-qemu\\x2d3\\x2drootdefault.scope", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 51
newfstatat(51, "", {st_mode=S_IFDIR|0755, st_size=0, ...}, AT_EMPTY_PATH) = 0
getdents64(51, 0x565141e6fa80 /* 54 entries */, 32768) = 1960
openat(AT_FDCWD, "/sys/fs/cgroup/machine.slice/machine-qemu\\x2d3\\x2drootdefault.scope/libvirt/cgroup.procs", O_RDONLY|O_CLOEXEC) = 54
newfstatat(54, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(54, "1507\n", 4096)                = 5
openat(AT_FDCWD, "/proc/1507/stat", O_RDONLY|O_CLOEXEC) = 57
newfstatat(57, "", {st_mode=S_IFREG|0444, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(57, "1507 (qemu-system-x86) S 1 1506 "..., 1024) = 338
ioctl(57, TCGETS, 0x7fff81c78510)       = -1 ENOTTY (Inappropriate ioctl for device)
read(57, "", 1024)                      = 0
close(57)                               = 0
openat(AT_FDCWD, "/proc/1507/cmdline", O_RDONLY|O_NOCTTY|O_CLOEXEC) = 57
newfstatat(57, "", {st_mode=S_IFREG|0444, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(57, "/usr/bin/qemu-system-x86_64\0-nam"..., 4104) = 2671
close(57)                               = 0
read(54, "", 4096)                      = 0
openat(AT_FDCWD, "/sys/fs/cgroup/machine.slice/machine-qemu\\x2d3\\x2drootdefault.scope/libvirt", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 57
newfstatat(57, "", {st_mode=S_IFDIR|0755, st_size=0, ...}, AT_EMPTY_PATH) = 0
getdents64(57, 0x565141e8b9b0 /* 53 entries */, 32768) = 1928
openat(AT_FDCWD, "/sys/fs/cgroup/machine.slice/machine-qemu\\x2d3\\x2drootdefault.scope/libvirt/vcpu1/cgroup.procs", O_RDONLY|O_CLOEXEC) = 58
newfstatat(58, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
read(58, 0x565141e8a430, 4096)          = -1 EOPNOTSUPP (Operation not supported)
```

**Comment by yuwata** [comment]
Hopefully fixed by #22095.
